Add a regional sea-level projection capability to MALI

hollyhan opened the pull request:
This PR adds a 1D gravitationally consistent sea-level model (Kendall et al., 2005; Gomez et al., 2010; Han et al., 2022; https://github.com/GomezGroup/1DSeaLevelModel_FWTW) as a submodule within MALI, adding the capability of regional sea-level projection associated with ice sheet changes solved by MALI. In this new addition, the sea-level model computes changes in the height of sea surface equipotential and viscoelastic solid Earth associated with surface (ice and ocean) loading changes, enabling MALI to take into account gravitational, rotational and deformational effects of the loading changes. 

The implemented requirements and modified/added codes in the PR are as follow: 

1. MALI to optionally build SLM from the top level directory (mpas-albany-landice):  modified `Makefile` in MALI's top level directory such that `SLM=True` in make command builds the SLM.
2. MALI to initialize SLM at t=0 of a simulation:  added wrapper subroutine 'slmodel_init' for initializing and the SLM in `mpas_li_bedtopo.F` in which sea-level model sub-modules are called.
3. MALI to call SL solver at every coupling time interval:  added wrapper subroutine 'slmodel_solve' for calling the SLM in `mpas_li_bedtopo.F` in which sea-level model sub-modules are called.
4. MALI to call the SLM at every coupling time interval (i.e., SLM timestep): 1) Modified `Registry.xml` to include a namelist value of the coupling time, 2) modified `mpas_li_core.F` to add an alarm to the land ice core clock object for the coupling time interval, and 3) modified `mpas_li_bedtopo.F` to use `mpas_timekeeping` and check/reset the alarm for the coupling time interval. 
5. MALI needs to accommodate SLM running on a single processor while running on multiple processors: Implement MPI scatter, gather and halo updates in `mpas_li_bedtopo.F`. MALI data (ice thickness) is gathered on the head node after which the SLM is called. 
6. Interpolation between MALI and SLM native grids should be done: Scripfiles `mpas_to_grid.nc` and `grid_to_mpas.nc` for the regional, unstructured MALI mesh and the global Gaussian grid are generated using MPAS-mesh tool `from_mpas.py` and `ncremap`, respectively. Subroutines `interpolate_init` and `interpolate` and MPI scatter/gather functions are copied from the ocean SAL PR#4472 [https://github.com/E3SM-Project/E3SM/pull/4472]
7. MALI and SLM must exchange model data: changes on the head node, MALI provides thickness as input to SLM through SLM subroutine `sl_solver` that calculates and outputs total bedtopography, which is used to update bedrock topography in `mpas_li_bedtopo.F`
8. Include the SLM namelist file such that SLM parameters can be accessed and modifed in MALI top-level directory and that the SLM does not need to be recompiled the parameters are modified.

matthewhoffman commented:
@hollyhan , thanks for opening this PR!  Can you update the PR title to something more descriptive, and expand the description to be more detailed?  For a major new functionality like this PR, we want the PR description to first give a summary of the overall capability, then describe the implementation, which subroutine were modified or added, and any important algorithmic details (e.g. copying the remapping and gather/scatter code from the ocean SAL PR - you can include a link to that PR).  I'll work on a more detailed line by line review starting today, but I thought I would mention that before I got started.

matthewhoffman reviewed (changes requested):
@hollyhan , overall this PR looks great and represents a lot of work!  I made a lot of comments, but they are all little things - many of which I should have instructed you in before you started working. :) A few of the items might require some discussion, but most should be self explanatory.  After the initial clean up pass, I'll take a closer look at a few details and might have more changes to request.  

The procedure for making changes is to add new commits that fix the various issues.  For the formatting and little changes, feel free to put many or all of the changes into a single commit.  For more significant changes, keep them in a separate commit.  After you've taken care of everything you can, push your updated branch to Github, and the PR will update automatically with your revisions.  When you are ready for me to take another look, request another review.

hollyhan commented on components/mpas-framework/Makefile:
```diff
@@ -335,11 +335,11 @@ gfortran:
 	"CXX_SERIAL = g++" \
 	"FFLAGS_PROMOTION = -fdefault-real-8 -fdefault-double-8" \
-	"FFLAGS_OPT = -O3 -m64 -ffree-line-length-none -fconvert=big-endian -ffree-form" \
+	"FFLAGS_OPT = -O3 -m64 -ffree-line-length-none -fconvert=big-endian -ffree-form -ffpe-summary=none" \
```
This is something we cherry-picked from Mark Peterson's PR to the E3SM repo. I forget for what this was for - I *think* it had something to do with floating point error that came up from the sea-level model.

matthewhoffman replied:
Oh, right - thanks for the reminder.  We will probably want to remove this commit from the PR before we merge it.  We can deal with that when the time comes.

hollyhan commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -273,10 +324,514 @@ subroutine li_bedtopo_finalize(domain, err)
+         ! interpolate thickness to Gaussian grid
+         call interpolate(toColValues, toRowValues, toSvalues, globalArrayIce, thickness_slGrid1D)
+			call interpolate(toColValues, toRowValues, toSvalues, mask_mesh, mask_slGrid1D)
+         call write_txt(thickness_slGrid1D, 'mali_iceload','')
```
it's not needed. Deleted it.

hollyhan commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -273,10 +324,514 @@ subroutine li_bedtopo_finalize(domain, err)
+!-----------------------------------------------------------------------
+
+   subroutine slmodel_solve(slmTimeStep, domain)
+      use li_advection
```
Is it not needed when calling `li_update_geometry`?

hollyhan replied:
Right, that's a great suggestion. So in changing this line, I also changed the original line of code inside of subroutine `li_bedtopo_solve`.

hollyhan commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -213,26 +242,48 @@ subroutine li_bedtopo_solve(domain, err)
+	        ! do nothing for now, but could calculate uplift rate here later instead.
+	      endif
+		else
+         call mpas_log_write("'sealevelmodel' should be selected for option 'config_uplift_method'", MPAS_LOG_ERR)
```
Wouldn't it be triggered only when the code is built with `SLM=True` but `config_uplift_method='sealevelmodel'`? I put this `if` statement intentionally to function as a prevention for user from building the SLM without selecting the right configuration choice for the uplift method. If you think this won't be useful, I will get rid of the if loop entirely.

hollyhan replied:
I actually got rid of the line in resolving the comments above (about including #ifdef within the check of config_uplift_method).

hollyhan commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -213,26 +242,48 @@ subroutine li_bedtopo_solve(domain, err)
          call mpas_log_write("Unknown option selected for 'config_uplift_method'", MPAS_LOG_ERR)
       endif
 
+#ifdef USE_SEALEVELMODEL
```
I made the changes, but it doesn't seem to be working - the error message does not get printed and the SLM runs all fine. Making the changes now, but I'll have to revisit this and maybe look at the code with you.

hollyhan replied:
We looked at this problem together and figured it out. I needed to do `make clean` whenever I compile the code with new compile flags.

hollyhan commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -273,10 +324,514 @@ subroutine li_bedtopo_finalize(domain, err)
+      integer :: unit_num
+      integer :: slmTimeStep, itersl_sh, dtime_sh !get these values from namelist
+      real    :: starttime_sh
+
```
I made the changes, but it doesn't seem to be working - the error message does not get printed and the SLM runs all fine. Making the changes now, but I'll have to revisit this and maybe look at the code with you.

hollyhan commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -273,10 +324,514 @@ subroutine li_bedtopo_finalize(domain, err)
+      call MPI_COMM_SIZE( domain % dminfo % comm, nProcs, err)
+
+      
+      ! perform the initialization on the head processor
```
That's a good point - thanks for pointing it out!

hollyhan commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -273,10 +324,514 @@ subroutine li_bedtopo_finalize(domain, err)
+         call interpolate(toColValues, toRowValues, toSvalues, globalArrayBedTopo, bedrock_slGrid1D)
+			call interpolate(toColValues, toRowValues, toSvalues, mask_mesh, mask_slGrid1D)
+
+         ! reformat the interpolated data
```
This is a great point! Made the suggested changes.

hollyhan replied:
This comment has been taken back in the following round of review. Changes have been made back to the original state.

hollyhan commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -273,10 +324,514 @@ subroutine li_bedtopo_finalize(domain, err)
+      real    :: starttime_sh
+
+		err_tmp = 0
+      itersl_sh = 1      ! * note: hardcode itersl_sh == 1.
```
We want to utilize the namelist for prescribing a value to this variable. It's hard coded now because I haven't worked on the namelist yet.

hollyhan replied:
This comment can be resolved as I now have the sea-level namelist working and thus removed Line 551

hollyhan commented:
Thank you for the very insightful comments, @matthewhoffman ! I have resolved most of the comments and updated the PR.

matthewhoffman reviewed (changes requested):
@hollyhan , thanks for all your hard work on revising this PR last week!  It is much easier to read and things look a lot more mature.  After a second reading, I've found a smaller number of additional changes.  I think with another round or two we will be close to done!

hollyhan commented on components/mpas-albany-landice/src/Registry.xml:
```diff
@@ -131,8 +131,20 @@
+        />
+                <nml_option name="config_slm_coupling_interval" type="character" default_value="0002-00-00_00:00:00" units="time"
+                            description="Time interval at which the sea-level model is called by MALI"
+                            possible_values="Any time interval of the format 'YYYY-MM-DD_HH:MM:SS'"
```
Right, yes. I've added the note in Registry.

hollyhan commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -213,26 +248,42 @@ subroutine li_bedtopo_solve(domain, err)
+			   call mpas_reset_clock_alarm(domain % clock, 'slmCouplingInterval', ierr=err_tmp)
+			   err = ior(err, err_tmp)
+
+			else
```
It's really weird, the spacing looks all consistent and correct on my text editor, and the tabs are replaced with three white spaces in the previous round of correction. I think a blank line between `err=ior(err, err_tmp)` and `else` is causing this weird indentation. I've removed the blank line, and we'll see how it appears in the updated PR.

hollyhan commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -273,10 +324,531 @@ subroutine li_bedtopo_finalize(domain, err)
+
+      integer :: err_tmp
+      integer :: unit_num
+      integer :: slmTimeStep, itersl_sh, dtime_sh !get these values from namelist
```
Right. It's been removed.

hollyhan commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -273,10 +324,531 @@ subroutine li_bedtopo_finalize(domain, err)
+      integer :: itersl_sh, dtime_sh
+      real    :: starttime_sh
+
+		err = 0
```
There isn't indentation nor white trailing space in this section except a blank line between Lines 552 and 554. I am not sure how to get rid of that indentation in 554. Can we take a look at it together?

hollyhan commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -273,10 +324,531 @@ subroutine li_bedtopo_finalize(domain, err)
+
+      if (curProc.eq.0) then
+         allocate(globalArrayThickness(nCellsGlobal), gatheredArrayThickness(nCellsGlobal))
+         allocate(globalArrayBedTopography(nCellsGlobal), gatheredArrayBedTopography(nCellsGlobal))
```
Right. Thanks for catching that!!

hollyhan commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -273,10 +324,531 @@ subroutine li_bedtopo_finalize(domain, err)
+      err = ior(err, err_tmp)
+
+      ! update bedTopography
+      bedTopography(:) = bedTopography(:) - upliftDiff(:)
```
`slChange_slGrid1D` represents the change in sea level (i.e., `delta(SL)=SL(j) - SL(j-1)`, where 'j' is a time index). This is equivalent to the negative of change in topography `delta(SL) = -delta(T) = T(j-1) - T(j)`, since `SL = -T`.  So we want to get the updated bedtopography` T(j)` by subtracting the sea-level change from the previous bedtopography `T(j-1)`. Hence, `T(j) = T(j-1) - delta(SL)`.

hollyhan replied:
We talked about change the variable name `upliftDiff` to `topoChange`, multiply `slChange_slGrid1D` by -1 and change the `-` sign in Line 621 to `+`

hollyhan commented:
Thank you so much for the second round of review, @matthewhoffman!
I've addressed all of the points of the conversation - I think it's ready for the next round. :)

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_core.F:
```diff
@@ -383,7 +390,6 @@ function li_core_run(domain) result(err)
 
       integer :: err, err_tmp, err_tmp2, globalErr
 
-
```
I'm still seeing about dozen whitespace changes in this file.  You might need to go to https://github.com/MALI-Dev/E3SM/pull/21/files to see what I mean.

matthewhoffman commented on components/mpas-albany-landice/src/Registry.xml:
```diff
@@ -1245,6 +1257,9 @@ is the value of that variable from the *previous* time level!
                      units="m s^{-1}"  description="time rate of change in bedTopography.  This field is prescribed as input and not currently calculated in the model."
                 />
+                <var name="TopoChange"            type="real"     dimensions="nCells Time"
+                     units="m"  description="amount of change in bedTopography.  This field is calculated by the sea-level model over its time step."
```
```suggestion
                     units="m"  description="amount of change in bedTopography.  This field is calculated by the sea-level model over its time step.  Positive values indicate bed uplift relative to sea level."
```

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -192,8 +227,17 @@ subroutine li_bedtopo_solve(domain, err)
+#endif
       elseif (trim(config_uplift_method)=='data') then
-
+#ifdef USE_SEALEVELMODEL
+         call mpas_log_write("MALI is built with 'SLM=true. Set 'config_uplift_method=sealevelmodel'", &
+	           MPAS_LOG_ERR)
+	      err = ior(err,1)
+#endif
```
Remove these lines too for same reason.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -213,26 +257,41 @@ subroutine li_bedtopo_solve(domain, err)
       endif
 
 
-
-
-
```
Restore these whitespace changes.

hollyhan replied:
I think the code actually looks correct with these lines removed?  The attached screenshot is the current status of the code without the whitespace changes.
![Screen Shot 2022-02-15 at 8 35 55 PM](https://user-images.githubusercontent.com/33441196/154191398-d5cd858e-f115-46df-aaa2-cbf1a965e627.png)

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -213,26 +257,41 @@ subroutine li_bedtopo_solve(domain, err)
       endif
 
    !--------------------------------------------------------------------
-
```
Remove this whitespace change.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -273,10 +332,538 @@ subroutine li_bedtopo_finalize(domain, err)
+      err_tmp = 0
+
+      ! initialize interpolation
+      call interpolate_init(domain, err)
```
Replace `err` with `err_tmp` in this call and add `err = ior(err, err_tmp)` after this line.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -273,10 +332,538 @@ subroutine li_bedtopo_finalize(domain, err)
+
+      ! perform the initialization on the head processor
+      if (curProc.eq.0) then
+         allocate(nCellsPerProc(nProcs))
```
This line reminded me that we should also add a new `li_bedtopo_finalize` routine that deallocates these arrays (and the other allocatable arrays in this routine that remain allocated for the entire simulation) before the model terminates.  It would need to be called from the MALI driver's finalize routine and it would need to check which config option is used and include ifdefs for if the sea level model is compiled.  I can explain further if it's hard to figure out the details!

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -273,10 +332,538 @@ subroutine li_bedtopo_finalize(domain, err)
+         call sl_call_readnl
+         call sl_solver_checkpoint(itersl, dtime)
+         call sl_timewindow(slmTimeStep)
+         call sl_solver(itersl, slmTimeStep, dtime, starttime, ismIceload, ismMask, slmSLchange)
```
One thing to think about - does the SLM have an error checks or error codes?  It would be useful to return an error if anything went wrong in the SLM.  That is probably a big task, so we don't need to include it in this PR, but could be a future improvement.

hollyhan replied:
The SLM does not have error checks nor codes. That would be a good thing to include in long term.

matthewhoffman replied:
Ok, let's not worry about that for this PR.  Before we get into full production runs, we will probably want to include some basic error checking from the SLM though.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -273,10 +332,538 @@ subroutine li_bedtopo_finalize(domain, err)
+         deallocate(meshMask)
+      endif
+
+      ! Perform Halo exchange update
+      call mpas_dmpar_field_halo_exch(domain,'bedTopography')
```
Move the halo update before calling `li_update_geometry` (to line 645).

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -273,10 +332,538 @@ subroutine li_bedtopo_finalize(domain, err)
+         call check( nf90_get_var(fromNcId, fromColId, fromColValues(:) ), err_tmp)
+         call check( nf90_get_var(fromNcId, fromRowId, fromRowValues(:) ), err_tmp)
+         call check( nf90_get_var(fromNcId, fromSId, fromSValues(:) ), err_tmp)
+
```
Add `err = ior(err, err_tmp)` here after all these checks.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -273,10 +332,538 @@ subroutine li_bedtopo_finalize(domain, err)
+         call check( nf90_inquire_dimension(fromNcId, nGridDimId, nGridName, nGrid), err_tmp)
+
+         ! Allocate matrices to read data into
+         allocate ( toRowValues (toNsLen) )
```
We'll have to remember to deallocate these in the finalize routine we add.

matthewhoffman reviewed (changes requested):
@hollyhan , this is looking really good!  There are just a handful of small things in this round.  The one bigger thing is I realized we should add a finalize routine that deallocates all the allocatable arrays.  Let me know if you info on the details of how to do that.  Otherwise, I think we are very close being done!  We'll need to have another discussion about the SLM repo and make sure all the details about that are sorted out before we can merge.

hollyhan commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -213,26 +257,41 @@ subroutine li_bedtopo_solve(domain, err)
       ! === error check
       if (err > 0) then
-          call mpas_log_write("An error has occurred in li_bedtopo_solve.", MPAS_LOG_ERR)
+          call mpas_log_write("An error has occurred in li_bedtopo_solve.")
```
Oops, I am not sure why I deleted that part. My bad.

hollyhan commented:
Thanks for the another round of review, @matthewhoffman! Most of the suggestions were easy to address. One thing I would need to discuss in more detail with you is adding the new `li_bedtopo_finalize` routine. I'll reach out to you!

hollyhan commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -232,7 +289,6 @@ end subroutine li_bedtopo_solve
 
 
 
-
```
I deleted this line to keep consistency with other routines that leave three white lines between the end of one routing and the beginning of another routine.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -267,16 +323,585 @@ subroutine li_bedtopo_finalize(domain, err)
+         call sl_deallocate_array
+
+         deallocate(globalArrayThickness)
+			deallocate(gatheredArrayThickness)
```
fix indentation here.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -267,16 +323,585 @@ subroutine li_bedtopo_finalize(domain, err)
+		if (trim(config_uplift_method)=='sealevelmodel') then
+	      if (curProc.eq.0) then
+	         deallocate(nCellsPerProc)
+			   deallocate(nCellsDisplacement)
```
fix indentation here.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -267,16 +323,585 @@ subroutine li_bedtopo_finalize(domain, err)
       err = 0
 
+#ifdef USE_SEALEVELMODEL
+      call mpas_pool_get_config(liConfigs, 'config_uplift_method', config_uplift_method)
```
Fix indentation throughout this routine - 3 spaces for each level.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -267,16 +323,585 @@ subroutine li_bedtopo_finalize(domain, err)
+         ! interpolate sea-level change from GL grid to MALI mesh.
+			! note: in the static sea-level theory, sea level and topography are globally defined !>
+			! and negative of each other. That is, topography change is negative of sea-level change
+         call interpolate(fromColValues, fromRowValues, fromSValues, -1*(slChangeSLgrid1D), globalArrayTopoChange)
```
```suggestion
         call interpolate(fromColValues, fromRowValues, fromSValues, -1.0_RKIND*(slChangeSLgrid1D), globalArrayTopoChange)
```

matthewhoffman reviewed (changes requested):
@hollyhan , I just have a couple small changes this time around.  After that, because this branch has been around for so long, let's rebase it on the current version of MALI-Dev/develop and then also remove commit a2771ad9fe, which won't be needed after the rebase.  You can do both operations at once with an interactive rebase.  I'm happy to walk you through it if needed.

hollyhan commented:
Thanks for the another round of review, @matthewhoffman! I've checked indentations and white trailing spaces.  Hopefully the code is ready to be merged after we update the SLM submodule.

hollyhan commented:
@matthewhoffman, the PR is now ready to be re-reviewed with the updated version of the SLM submodule!

matthewhoffman commented:
@hollyhan , this looks good.  I only see one issue.  Even after rebasing, there are changes to `components/mpas-framework/Makefile`, which we need to remove from your branch.  They occur in fdaccc2153.  The best way to deal with that is the following:
1. `git rebase -i a9cc375021`
2. In the editor window that appears, change fdaccc2153 to `edit` 
3. When you get the command prompt for editing that commit, do: `git checkout HEAD^ -- components/mpas-framework/Makefile`  This resets the version of the Makefile to its state in the previous commit (i.e. before it was modified by this commit)
4. `git commit --amend`  , modify the commit message to exclude the part about Makefile changes
5. `git rebase --continue`

Once that completes, check things worked by doing `git log components/mpas-framework/Makefile`.  The first commit that appears should be from Feb 1, unrelated to your branch (767728dfc21).  Also, do a normal `git logg` and you should see your 36 commits as you do now.  If that looks good, force push your branch again.

hollyhan commented:
Thanks for the rebase instruction, @matthewhoffman. I've followed the steps and updated the PR.

matthewhoffman commented:
@hollyhan , thanks for updating that.  I don't see any other issues, so we're almost done!  I'd like to spend some time doing some detailed testing actually running the code this week, and I'd also like @trhille to do his own cursory review (Trevor, you probably saw I requested a review from you).  Holly, can you add a list of the input files you have been testing with and where I/we can find them on Badger or Cori?

hollyhan commented:
@matthewhoffman and @trhille, I've set up the SLM input files and folders you'll need for testing coupled MALI-SL simulations on Badger `/lustre/scratch4/turquoise/hollyhan/FILES_FOR_TESTING_MALI/`. Within this directory, you'll see three folders `INPUT_FILES`, `ICELOAD_SLM`,  and `OUTPUT_SLM` - the first folder has all of the input files you'll need and the two latter folders are where the SLM will store its output files (at least based on the current `namelist.sealevel` file. In the directory, I've also put other files that you'll need in your run folder - `mpas_to_grid.nc` and `grid_to_mpas.nc` are the mapfiles for interpolation between the MALI mesh and the SLM grid.

trhille commented:
> This PR adds a 1D gravitationally consistent sea-level model (Kendall et al., 2005; Gomez et al., 2010; Han et al. 2022, [https://github.com/hollyhan/SeaLevelModel](url)) as a submodule within MALI
Note that this github link does not work.

trhille commented:
I'm getting an error when I try to compile this on Badger. I'm using:
`source /users/mhoffman/setup_badger_mods.2021-03-09.sh
make gfortran ALBANY=true DEBUG=true SLM=true`

And I get:
"""
mpas_li_bedtopo.F:399:25:

mpas_li_bedtopo.F:366:10:

       use sl_io_mod                       !< this is part of the SLM code
          2               
mpas_li_bedtopo.F:399:25:

       integer :: unit_num      ! SLM variable
                         1
Error: Symbol ‘unit_num’ at (1) conflicts with symbol from module ‘user_specs_mod’, use-associated at (2)
"""

trhille commented:
@hollyhan I am getting an upper bound error when I try to run this:
```
At line 728 of file mpas_li_bedtopo.F
Fortran runtime error: Index '1627' of dimension 1 of array 'rowvalues' above upper bound of 1626
```
I'm not at all sure that I'm doing this correctly, however. I copied your directory  `/lustre/scratch4/turquoise/hollyhan/FILES_FOR_TESTING_MALI/` to my scratch dir, added `streams.landice` and `namelist.landice` and ran on 8 procs. I have only some basic physics enabled:
```
config_velocity_solver = 'none'
config_thickness_advection = 'fo'
config_tracer_advection = 'fo'
config_calving = 'none'
config_restore_calving_front = .true.
config_thermal_solver = 'none'
config_basal_mass_bal_float = 'temperature_profile'
config_uplift_method='sealevelmodel'
```

trhille commented:
Oh, I misinterpreted what was ready to be tested and I compiled with `DEBUG=true`. I'll test again with `DEBUG=false`.

trhille reviewed (changes requested):
It seems that the variable `lot` (number of data vectors) is not defined, or is defined incorrectly, leading to that upper bound error I saw arising in L1720 of spharmt.f90. We should define this variable somewhere. Since this is a bug in legacy code and not in your coupling, this could be a separate PR if needed.

trhille commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -267,16 +323,585 @@ subroutine li_bedtopo_finalize(domain, err)
+!>
+!-----------------------------------------------------------------------
+
+   subroutine interpolate(colValues, rowValues, sValues, dataIn, dataOut)
```
"interpolate" is a pretty generic name. Is there a more specific name that could be used?

hollyhan replied:
More specific name could be something like `interpolate_between_MALImesh_SLMgrid`, but this is too long, and also we are adopting (i.e., copying and pasting) the subroutine that MPAS-O [https://github.com/E3SM-Project/E3SM/blob/4523e80e542f5e46952365257fb2f2e38de38039/components/mpas-ocean/src/shared/mpas_ocn_vel_self_attraction_loading.F#L964-L986](url), so I'd prefer to leave the subroutine as is. Having a generic name for the interpolation won't cause us too much confusion in running MALI  because this `interpolation`subroutine is used only in coupled MALI-SLM simulations.

trhille commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -267,16 +323,585 @@ subroutine li_bedtopo_finalize(domain, err)
+      err = ior(err, err_tmp)
+
+      ! update bedTopography
+      bedTopography(:) = bedTopography(:) + topoChange(:)
```
It would probably be useful to update `upliftRate` here as well, even though it isn't used in this routine. I would also slightly prefer the variable name `bedTopographyChange` over `topoChange`, since the ice sheet surface is also topography.

hollyhan replied:
@trhille, I've renamed the variable `topoChange` to `bedTopographyChange` - thanks for bringing up the good point! 
In regards to updating `upliftRate` - it seems irrelevant at least for now to include the variable in the submodule when it's not calculated/used at all in coupled simulations. @matthewhoffman and I shortly discussed before and we left it at not using the `upliftRate`, but maybe we can resume the discussion and revisit our decision. I'll bring this up next time we meet. Until we make a collective decision, I'll leave the code as is.

trhille reviewed (changes requested):
The coupling interval is set in both MALI and SLM settings, and there is no check on consistency.  This leads to an inconsistency in the model time between MALI and SLM if they are not set to the same value.

trhille commented:
I tested this on Badger on 8 processors and found some strange behavior.
start time:
![image](https://user-images.githubusercontent.com/17446278/160475904-79f8490c-5f2f-4db9-875c-4a191c4952ad.png)

end time:
![image](https://user-images.githubusercontent.com/17446278/160475935-1fb788e4-2d9c-4fa4-a602-94963a039859.png)

The white contours here are cellProcID. I don't _think_ this is caused by a missing halo update, because in my experience that usually causes artifacts that align with the cellProcID contours. But with the multiple interpolations and switching between 1 proc for the sea-level model and 8 procs for MALI, I'm not sure how that would manifest itself here.

Note that changes here are driven by _very_ negative imposed SMB, which causes about 1km of thinning in 10 years.

hollyhan commented:
Thanks so much for testing this, @trhille!
I have resolved the issues you are seeing in `bedTopography` in the latest commit. Apparently, I have accidentally assigned a wrong mapping file for the interpolation from SLM grid to MALI mesh. This wrong assignment was introduced in the last few commits, and I must have committed unwanted changes while I was trying to figure out the boundcheck error we are seeing in the subroutine `interpolation_init`. Testing the code with the new fix, I am seeing the results that show evolving `bedTopography`, for both runs on 1 proc and  8 procs.
[Top panel: `bedTopography` at simulation year 50 on 8 procs]
[Middle panel: `bedTopography` at simulation year 0 on 1 proc]
[Bottom panel: `bedTopography` at simulation year 50 on 1 proc]
![Screen Shot 2022-03-30 at 1 43 43 PM](https://user-images.githubusercontent.com/33441196/160917831-bbbd0f3a-ce06-48e9-97be-fd2614f3fe1a.png)
![Screen Shot 2022-03-30 at 1 45 30 PM](https://user-images.githubusercontent.com/33441196/160918148-d40cfa89-8dc7-433b-a97c-d948cc237ec8.png)
![Screen Shot 2022-03-30 at 1 46 15 PM](https://user-images.githubusercontent.com/33441196/160918283-bd688ce7-bca1-400b-a9ff-ac6baa228ec1.png)

Hopefully you'll get the same results in the next testing!

trhille commented:
@hollyhan, are these panel descriptions correct?
 [Top panel: `bedTopography` at simulation year 50 on 8 procs] [Middle panel: `bedTopography` at simulation year 0 on 1 proc] [Bottom panel: `bedTopography` at simulation year 50 on 1 proc] ![Screen Shot 2022-03-30 at 1 43 43 PM]

hollyhan commented:
@trhille , yes they are - they are just in different colour scale

trhille commented:
@hollyhan, can you explain how to set the length of the simulation? I'm using these settings in `namelist.landice`:
```   
config_start_time = 2007-07-01_00:00:00
config_stop_time = '2107-07-01_00:00:00'
```
and these in `namelist.sealevel`:
```
&time_config
    itersl = 1
    starttime = 0
    dt1 = 5
...
&timewindow_config
    L_sim = 100
    dt1 = 5
    dt2 = 10
    dt3 = 10
    dt4 = 10
    Ldt1 = 20
    Ldt2 = 0
    Ldt3 = 0
    Ldt4 = 0
```

It ran for 55 years and then ended with:
```
At line 288 of file sl_init_mod.f90 (unit = 201)
Fortran runtime error: Cannot open file 'INPUT_FILES/icemodel/iceload5g11': No such file or directory
```
Does the number of `iceload5g` files limit the length of the run? And if so, how do I set up a run of arbitrary duration?

trhille commented:
From my 55-yr runs, I can confirm that the issues with noisy or unstable `bedTopography` have been resolved by [4319b7b](https://github.com/MALI-Dev/E3SM/pull/21/commits/4319b7b45ce7b66e5e4ec5ab4d798acd7a99ade2), that bedTopography is evolving, and that bedTopography is bit-for-bit between simulations on 1 processor and 8 processors.

Runs started in 2007. left: 1 processor; middle 8 processors; right: difference between the two:
![image](https://user-images.githubusercontent.com/17446278/160940554-8234aaee-8f6b-44c7-a6a4-64fc6ba8fbaa.png)

hollyhan commented:
> Need to clean `src/SeaLevelModel` directory in these lines: https://github.com/MALI-Dev/E3SM/blob/hollyhan/mali/add_SLmodel/components/mpas-albany-landice/src/Makefile#L54-L63

@trhille, I realize that @matthewhoffman had previously made `make clean` work for `src/SeaLevelModel' with `make clean SLM=true`.  I've modified the makefile slightly so that simply doing `make clean` will also clean files in `src/SeaLevelModel` if the SLM submodule exists.

hollyhan commented:
> It seems that the variable `lot` (number of data vectors) is not defined, or is defined incorrectly, leading to that upper bound error I saw arising in L1720 of spharmt.f90. We should define this variable somewhere. Since this is a bug in legacy code and not in your coupling, this could be a separate PR if needed.

@trhille - thanks for catching this. This particular piece of code was adopted by the group that developed original form of the sea-level model that we are using. Leaving it as is has been no problem so far because the SLM has been compiled without ` -fbounds-check` and its results were benchmarked to the ancestral version of sea-level model.  I do also think that we need to address this issue at some point and apply the same debug flags as we do in compiling MALI for compiling the SLM. Since this code belongs to the SLM submodule, we'll make a PR directly in the SLM repository and update the submodule here once the issue has been addressed.

hollyhan commented:
> The coupling interval is set in both MALI and SLM settings, and there is no check on consistency. This leads to an inconsistency in the model time between MALI and SLM if they are not set to the same value.

Thanks for noting this, @trhille. I've added a check to the code that checks the consistency between coupling interval setup by MALI and SLM.

hollyhan commented:
> @hollyhan, can you explain how to set the length of the simulation? I'm using these settings in `namelist.landice`:
> 
> ```
> config_start_time = 2007-07-01_00:00:00
> config_stop_time = '2107-07-01_00:00:00'
> ```
> 
> and these in `namelist.sealevel`:
> 
> ```
> &time_config
>     itersl = 1
>     starttime = 0
>     dt1 = 5
> ...
> &timewindow_config
>     L_sim = 100
>     dt1 = 5
>     dt2 = 10
>     dt3 = 10
>     dt4 = 10
>     Ldt1 = 20
>     Ldt2 = 0
>     Ldt3 = 0
>     Ldt4 = 0
> ```
> 
> It ran for 55 years and then ended with:
> 
> ```
> At line 288 of file sl_init_mod.f90 (unit = 201)
> Fortran runtime error: Cannot open file 'INPUT_FILES/icemodel/iceload5g11': No such file or directory
> ```
> 
> Does the number of `iceload5g` files limit the length of the run? And if so, how do I set up a run of arbitrary duration?

How you have defined the length of simulation is "almost" correct - you would also want to set `Ldt1` same as the `L_sim` if you don't want to use the time window algorithm. I realize this is really not straight forward just by looking at the namelist because there's no description of the name of variables. You can get more detailed description of the sea level namelist variables here [(https://github.com/MALI-Dev/1DSeaLevelModel_FWTW/blob/main/user_specs_mod.f90#L134-L135)](url).

The number of iceload files (e.g., `iceload5g` in `inputfolder_ice` in `namelist.sealevel`) does limit the length of the run in an indirect way. Those input ice files are globally-defined ice thickness outside (they can be from other ice model results/reconstructions), and they need to be pre-processed/generated before the simulation starts. They get merged with ice thickness passed by MALI at every coupling timestep when the SLM is called (i.e., `iceload5g0` will get merged with ice thickness over the Thwaites region that MALI provides at the initial timestep, `iceload5g1` will get merged with MALI ice thickness at year 5 when the coupling interval is 5 years... `iceload5g10` will be merged with model year 50.) So when MALI was calling the SLM at year 55, the SLM is looking for `iceload5g11`.  I've only put 10 ice files in the inputfile for the test purposes, and that is why the run stopped. What you can do for now if you want to test the run for 100 years is to simply copy one of the input files like `iceload5g0` and make `iceload5g11...iceload5g20`.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -471,6 +472,17 @@ subroutine slmodel_init(domain, err)
+         ! First, check consistency in coupling interval set up in MALI and SLM
+         err = 0
+         call mpas_pool_get_config(liConfigs, 'config_slm_coupling_interval', config_slm_coupling_interval)
+         read(config_slm_coupling_interval(1:4),*) slm_coupling_interval
```
This looks like it is assuming the `config_slm_coupling_interval` option has zeros for month, day, hr, min, sec.  I think we might want the option to run with coupling intervals of smaller than a year to do sensitivity tests to the coupling interval.  We can use the MPAS timekeeping operations to parse the string more carefully.

matthewhoffman replied:
Given that the SLM just uses decimal years for timekeeping, it would be really tricky to ensure an exact match in the time interval (due to MPAS having the option to track leap years, MPAS keeping track of differing days per month, etc.).  I think a match in time interval within 1% would be a reasonable approximation to avoid potentially complicated accounting here.  

The most consistent way to do the calculation would be to use `mpas_set_timeInterval` and `mpas_get_timeInterval` in `components/mpas-framework/src/framework/mpas_timekeeping.F` to return the MPAS time interval in seconds (the `dt` output argument), and compare that to the SLM time interval.  Then throw an error if they differ by, say, >1%.  One complication there is that for time intervals that require months and/or years, `mpas_get_timeInterval` needs to know the start time, because the result will depend on the start time for many time intervals (e.g., 1 month).  For the purposes of this calculation, the start time is not important, but you'd have to provide something.  It might be best to hard code a time, like `2001-01-01_00:00:00` (I picked an arbitrary start of a year that isn't a leap year.).  That is probably better than using something like the model start time, which would result in possibly slightly different results depending on start year.

This ends up being rather complex for a simple comparison, but I don't see a good way around it given the two different timekeeping systems in the two models.

matthewhoffman commented:
@hollyhan , I have the morning free to finish off testing of this PR and hopefully get it merged.  My first test was to compile and run with `SLM=false` and ensure everything still works as expected without the SLM activated.  However, I see a bunch of compile errors where it looks like the build system is still looking for SLM code when I do that.  Can you see if you can reproduce that error, and if so, see if you can fix it?  I suspect there is some code not properly protected by the ifdefs.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -213,6 +250,26 @@ subroutine li_bedtopo_solve(domain, err)
+
+            call mpas_log_write("Calling the SLM. SLM timestep $i", intArgs=(/slmTimeStep/))
+            call slmodel_solve(slmTimeStep, domain)
+
```
```suggestion
call mpas_log_write("----- SLM completed timestep $i -----", intArgs=(/slmTimeStep/))
```
Let's add a completion message so it's really clear in the log file which parts of the file are being written by SLM.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -213,6 +250,26 @@ subroutine li_bedtopo_solve(domain, err)
+
+            slmTimeStep = slmTimeStep + 1
+
+            call mpas_log_write("Calling the SLM. SLM timestep $i", intArgs=(/slmTimeStep/))
```
```suggestion
            call mpas_log_write("----- Calling the SLM. SLM timestep $i -----", intArgs=(/slmTimeStep/))
```

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_bedtopo.F:
```diff
@@ -87,10 +114,17 @@ subroutine li_bedtopo_init(domain, err)
+      call mpas_pool_get_config(liConfigs, 'config_uplift_method', config_uplift_method)
+      if (trim(config_uplift_method)=='sealevelmodel') then
+         ! initialize the 1D sea-level model
+         call slmodel_init(domain, err)
```
Let's add clear before and after log messages around this call like I suggested for the `slmodel_solve` call below.

hollyhan commented:
> @hollyhan , when I compile and run in debug mode, I get the following error:
> 
> ```
> At line 1720 of file spharmt.f90
> Fortran runtime error: Index '513' of dimension 1 of array 'a' above upper bound of 512
> ```
> 
> If disable debug mode, the error goes away. Have you seen this?

@matthewhoffman , Yes - that's the error message from the module code `spharmt.f90` in SLM, which was adopted by the group in harvard when the new original ice-age version of the SLM was written. It's a bound error, and it doesn't affect the calculation, which is why I think no efforts were made to correct it by the original developers. Trevor ran into this issue during his check, and solved by `make clean`-ing the SLM directory (https://github.com/MALI-Dev/E3SM/pull/21#issuecomment-1080832694), which is now done in the top level directory (1469bf4d7737b8c66ce26a5a8e3a65b165aae1c5)

matthewhoffman commented:
I've tested the COMPASS `full_integration` suite in three ways, all with `ALBANY=true`:
1. `SLM=false DEBUG=true`
2. `SLM=true DEBUG=false` (setting DEBUG to false because of the error I get when both are true)
3. `SLM=true DEBUG=true` (none of existing tests include running the SLM, so the out of bounds error does not occur)
In all three cases, all existing tests pass.  None of the existing tests test the new functionality of course, but I can confirm that the changes do not affect any existing functionality.  

I also confirmed:
* the expected error message occurs if you try to compile with `SLM=true` but did not initialize the SLM git submodule.
* if you compile without SLM but try to run with it you get a useful error message

I will proceed to testing runs that actually use the SLM with the Thwaites configuration Holly set up.

trhille commented:
> ice-age version of the SLM was written. It's a bound error, and it doesn't affect the calculation, which is why I think no efforts were made to correct it by the original developers. Trevor ran into this issue during his check, and solved by `make clean`

@hollyhan That fix was when using `DEBUG=false`. So it seems that needs to be fixed for debug mode.

trhille reviewed (approved):
@hollyhan, _great_ work on this! After discussion with @matthewhoffman, I'm approving this PR with the knowledge that a few clean-up items will be addressed in another PR.

matthewhoffman reviewed (approved):
@hollyhan, I finished testing with the Thwaites mesh and see the same results as you and @trhille.  I agree this looks really good and works as desired.  It also nicely avoid affecting model behavior when the sea level model is not in use.  Thanks for all your hard work in all the rounds of revisions and unexpected issues that popped up.  We discussed the few clean-up items that we can leave for a follow up PR, but this is ready to merge!
